File: src/content/algorithms/counting-sort/code.py
def counting_sort(array):  # #@function-def
    if not array:
        return array

    max_val = max(array)  # #@find-max

    count = [0] * (max_val + 1)  # #@count-array

    for num in array:  # #@counting
        count[num] += 1

    for i in range(1, len(count)):  # #@cumulative
        count[i] += count[i - 1]

    output = [0] * len(array)  # #@output

    for i in range(len(array) - 1, -1, -1):  # #@placement
        output[count[array[i]] - 1] = array[i]
        count[array[i]] -= 1

    return output

File: src/content/algorithms/counting-sort/code.rs
pub fn counting_sort(arr: &[usize]) -> Vec<usize> {  // #@generic-function
    if arr.is_empty() {
        return vec![];
    }

    let max_val = *arr.iter().max().unwrap();  // #@find-max

    let mut count = vec![0usize; max_val + 1];  // #@count-array

    for &num in arr {  // #@counting
        count[num] += 1;
    }

    for i in 1..count.len() {  // #@cumulative
        count[i] += count[i - 1];
    }

    let mut output = vec![0usize; arr.len()];  // #@output

    for &num in arr.iter().rev() {  // #@placement
        output[count[num] - 1] = num;
        count[num] -= 1;
    }

    output
}
